File: src/lib/observability.ts
/**
 * Error monitoring, gated on SENTRY_DSN. When the DSN is unset, everything is a
 * logged no-op so dev/CI need no Sentry account. @sentry/node is imported lazily
 * so the no-DSN path pulls in no SDK code.
 */
type SentryModule = typeof import('@sentry/node');

let initialized = false;
let sentry: SentryModule | null = null;

export async function initSentry(): Promise<boolean> {
  if (initialized) return sentry !== null;
  initialized = true;

  const dsn = process.env.SENTRY_DSN;
  if (!dsn) return false;

  try {
    const Sentry = (await import('@sentry/node')) as SentryModule;
    Sentry.init({
      dsn,
      tracesSampleRate: Number(process.env.SENTRY_TRACES_SAMPLE_RATE) || 0.1,
      environment: process.env.VERCEL_ENV || process.env.NODE_ENV || 'development',
    });
    sentry = Sentry;
    return true;
  } catch (e) {
    console.error('[observability] Sentry init failed', e);
    return false;
  }
}

export async function captureError(err: unknown, context?: Record<string, unknown>): Promise<void> {
  if (!sentry) {
    if (!initialized) await initSentry();
  }
  if (sentry) {
    sentry.captureException(err, context ? { extra: context } : undefined);
  } else {
    console.error('[error]', err, context ?? '');
  }
}

File: scraper/observability.py
"""Error monitoring for the scraper daemon, gated on SENTRY_DSN.

No-op (logged) when SENTRY_DSN is unset or sentry_sdk is not installed, so the
daemon runs anywhere without an account.
"""

import os
import logging

log = logging.getLogger("peakr-observability")

_initialized = False
_sentry = None


def init_sentry() -> bool:
    """Initialize Sentry if configured. Returns True if active."""
    global _initialized, _sentry
    if _initialized:
        return _sentry is not None
    _initialized = True

    dsn = os.environ.get("SENTRY_DSN")
    if not dsn:
        return False
    try:
        import sentry_sdk

        sentry_sdk.init(
            dsn=dsn,
            traces_sample_rate=float(os.environ.get("SENTRY_TRACES_SAMPLE_RATE", "0.1")),
            environment=os.environ.get("SENTRY_ENV", "production"),
        )
        _sentry = sentry_sdk
        log.info("Sentry initialized")
        return True
    except Exception as e:  # ImportError or init failure
        log.error("Sentry init failed: %s", e)
        return False


def capture_exception(err: BaseException) -> None:
    """Report an exception to Sentry if active; otherwise log it."""
    if _sentry is not None:
        try:
            _sentry.capture_exception(err)
            return
        except Exception:
            pass
    log.error("Captured exception (Sentry inactive): %s", err)
